src/github.com/matrix-org/dendrite/clientapi/routing/whoami.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2018 Vector Creations Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package routing
+
+import (
+	"net/http"
+	"strings"
+
+	"github.com/matrix-org/dendrite/clientapi/auth"
+	"github.com/matrix-org/dendrite/clientapi/auth/storage/accounts"
+	"github.com/matrix-org/dendrite/clientapi/auth/storage/devices"
+	"github.com/matrix-org/dendrite/clientapi/jsonerror"
+	"github.com/matrix-org/dendrite/common/config"
+	"github.com/matrix-org/util"
+)
+
+type whoamiResponse struct {
+	UserID string `json:"user_id"`
+}
+
+// GetUserFromRequest extracts UserID from the given HTTP request,
+// Finds local user or an application service user,
+// On failure returns an error response which can be sent to the client.
+func GetUserFromRequest(
```

Comment:
it feels to me like a mistake to be doing this as a special-case for `whoami`. It's relevant to the majority of the C-S APIs.

Could we not roll it into `common.MakeAuthAPI` ?

Reply:
Thanks, 
this is indeed more practical. The correct place for this function should have been `clientapi/auth` but while doing so, I realised that `clientapi/auth/storage/devices/storage.go` imports `clientapi/auth`. This is a recipe for import cycles.
I am going to refactor a little bit and integrate this into `common.MakeAuthAPI`.
I hope it's alright.